pkg/cloud/vsphere/vsphere_config_transformer.go

```diff
@@ -99,3 +101,47 @@ func setVirtualCenters(cfg *ccmConfig.CPIConfig, vSphereSpec *configv1.VSpherePl
 		}
 	}
 }
+
+// setDualStack updates the configuration required by the cloud-provider-vsphere to explicitly set
+// value of IPFamilyPriority instead of using the default which is IPv4. This is needed by the
+// cloud provider in order to properly filter IP addresses that feed the instance metadata.
+//
+// We rely on the Service Networks configuration that initially comes from o/installer and later
+// from the Cluster Network Operator as those two components take care of validating that clusters
+// with dual-stack configuration have exactly 2 of them and that they match the required order.
+//
+// We are mangling with the ExcludeNetworkSubnetCIDR param here because VM agent by default detects
+// also IP addresses that are used by us internally and which should never be exposed as node IPs
+// (i.e. API VIP and Ingress VIP for IPI installations and fd69::2 which is internal to OVN-K8s).
+//
+// Ref.: https://issues.redhat.com/browse/OCPBUGS-18641
+func setDualStack(cfg *ccmConfig.CPIConfig, status *configv1.VSpherePlatformStatus, nodeNetworking *configv1.VSpherePlatformNodeNetworking, network *configv1.Network) {
+	if network != nil && len(network.Spec.ServiceNetwork) == 2 {
+		// Extensive validations are performed by o/installer so that here we already know that
+		// if the configuration is dual-stack, we will have exactly 2 service networks and if
+		// single-stack then 1 service network. Simplified logic here is applied to avoid code
+		// duplication.
+		//
+		// Ref.: https://github.com/openshift/installer/blob/6471b31/pkg/types/validation/installconfig.go#L241
+		if net.IsIPv4CIDRString(network.Spec.ServiceNetwork[0]) {
+			cfg.Global.IPFamilyPriority = []string{"ipv4", "ipv6"}
+		} else {
+			cfg.Global.IPFamilyPriority = []string{"ipv6", "ipv4"}
+		}
+
+		if status != nil {
+			for _, addr := range append(status.APIServerInternalIPs, status.IngressIPs...) {
+				if net.IsIPv4String(addr) {
+					addr = addr + "/32"
+				} else {
+					addr = addr + "/128"
+				}
+				nodeNetworking.External.ExcludeNetworkSubnetCIDR = append(nodeNetworking.External.ExcludeNetworkSubnetCIDR, addr)
+				nodeNetworking.Internal.ExcludeNetworkSubnetCIDR = append(nodeNetworking.Internal.ExcludeNetworkSubnetCIDR, addr)
+			}
+		}
+
+		nodeNetworking.External.ExcludeNetworkSubnetCIDR = append(nodeNetworking.External.ExcludeNetworkSubnetCIDR, "fd69::2/128")
```

Comment:
maybe a named constant for `fd69::2/128` so we can crossref with ovn-kubernetes.

Reply:
I'd totally agree if we had a common/shared library that ovn-k8s would consume. In the state like it is now, I'd say address is an address and it's not a big diff if there is variable or not (especially that I use it only in a single place)

Reply:
we avoid magic numbers?  and it is used twice.   but maybe we add a comment in OVN-k8s as well.

https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/pkg/config/config.go#L145

can customer override `V6HostMasqueradeIP`?

Reply:
Yep, it's configurable  https://github.com/ovn-org/ovn-kubernetes/pull/3594

Reply:
https://github.com/openshift/cluster-network-operator/pull/1807

Reply:
Err, we can pull from the API?  https://github.com/openshift/api/pull/1410

Reply:
I'm looking into this. It's not nice because they didn't add it as Network CRD but only in the Operator Config CRD so I'm not sure if I can access it easily from inside the 3cmo

Reply:
To clarify - until https://github.com/openshift/cluster-network-operator/pull/1807 merges, this value is **not configurable**. I am now discussing with OVN-K8s folks the rationale, but at this moment this value is the only one that we will see in the wild.

Having said this, I wonder if we can push this change as it is now and then when OVN-K8s folks are ready with https://github.com/openshift/cluster-network-operator/pull/1807 (hopefully with revisited API that stores the custom subnet in the Network CR) improve it for 4.15

Reply:
Yep, I agree we should merge this to unblock dual-stack.